From b008a49a555a64684f42f89a6e27f9fda5f4ce89 Mon Sep 17 00:00:00 2001
From: Christoph Reiter <reiter.christoph@gmail.com>
Date: Tue, 6 Jan 2026 12:11:05 +0100
Subject: [PATCH 011/N] build: Add PYD_PLATFORM_TAG

Add PYD_PLATFORM_TAG to distinguish C extensions built with different
MinGW toolchain configurations. The tag follows the format:
  mingw_<cpu_arch>_<c_runtime>_<toolchain>

Where:
- cpu_arch: x86_64, i686, aarch64, or armv7
- c_runtime: ucrt or msvcrt
- toolchain: gnu or llvm

This prevents loading extensions built with incompatible configurations.

Co-authored-by: Naveen M K <naveen521kk@gmail.com>
---
 Makefile.pre.in |  1 +
 configure.ac    | 55 +++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/Makefile.pre.in b/Makefile.pre.in
index 38a355a..f1d3451 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -173,6 +173,7 @@ CONFINCLUDEPY=	$(CONFINCLUDEDIR)/python$(LDVERSION)
 # Symbols used for using shared libraries
 SHLIB_SUFFIX=	@SHLIB_SUFFIX@
 EXT_SUFFIX=	@EXT_SUFFIX@
+PYD_PLATFORM_TAG = @PYD_PLATFORM_TAG@
 LDSHARED=	@LDSHARED@ $(PY_LDFLAGS)
 BLDSHARED=	@BLDSHARED@ $(PY_CORE_LDFLAGS)
 LDCXXSHARED=	@LDCXXSHARED@ $(PY_LDFLAGS)
diff --git a/configure.ac b/configure.ac
index dd56f0d..991806a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6475,6 +6475,61 @@ esac
 # check for endianness
 AC_C_BIGENDIAN
 
+AC_SUBST(PYD_PLATFORM_TAG)
+# Special case of PYD_PLATFORM_TAG with python build with mingw.
+# Python can with different cpu arch and c runtime as well as different
+# toolchain. We follow this`mingw_<cpu_arch>_<c_runtime>_<toolchain>`
+# convention for PYD_PLATFORM_TAG. Where:
+# `cpu_arch` = `x86_64`, `aarch64` or `i686`
+# `c_runtime` = `msvcrt` or `ucrt`
+# `toolchain` = `gnu` or `llvm`
+PYD_PLATFORM_TAG=""
+case $host in
+  *-*-mingw*)
+    # check if we are linking to ucrt
+    AC_MSG_CHECKING(whether linking to ucrt)
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE([[
+    #include <stdio.h>
+    #ifndef _UCRT
+        #error no ucrt
+    #endif
+    int main(){ return 0; }
+    ]])],[linking_to_ucrt=yes],[linking_to_ucrt=no])
+    AC_MSG_RESULT($linking_to_ucrt)
+    ;;
+esac
+case $host_os in
+    mingw*)
+  AC_MSG_CHECKING(PYD_PLATFORM_TAG)
+  PYD_PLATFORM_TAG="mingw"
+  case $host in
+  i686-*-mingw*)
+    PYD_PLATFORM_TAG+="_i686"
+    ;;
+  x86_64-*-mingw*)
+    PYD_PLATFORM_TAG+="_x86_64"
+    ;;
+  aarch64-*-mingw*)
+    PYD_PLATFORM_TAG+="_aarch64"
+    ;;
+  armv7-*-mingw*)
+    PYD_PLATFORM_TAG+="_armv7"
+    ;;
+  esac
+  if test $linking_to_ucrt = no; then
+    PYD_PLATFORM_TAG+="_msvcrt"
+  else
+    PYD_PLATFORM_TAG+="_ucrt"
+  fi
+  if test "$ac_cv_cc_name" = "clang"; then
+    # it is CLANG32
+    PYD_PLATFORM_TAG+="_llvm"
+  else
+    PYD_PLATFORM_TAG+="_gnu"
+  fi
+  AC_MSG_RESULT($PYD_PLATFORM_TAG)
+esac
+
 # ABI version string for Python extension modules.  This appears between the
 # periods in shared library file names, e.g. foo.<SOABI>.so.  It is calculated
 # from the following attributes which affect the ABI of this Python build (in
